code/modules/projectiles/guns/projectile.dm

```diff
@@ -46,6 +46,7 @@
 		return
 	else if(magazine.ammo_count())
 		chambered = magazine.get_round()
+		magazine.update_icon()
```

Comment:
update_icon is called to many times now down the calling tree. Example in automatic.dm at line 47.
Best to flush those out as well to avoid redundant calls

Reply:
I've moved the update_icon() but I don't know if this will fix the performance problem.

The reason for adding an update_icon() was because there’s a bug. The magazine overlay for the Bulldog isn't updated as rounds are pulled from the magazine into the weapon chamber.

If you have two rounds in the Bulldog, one in the chamber and one in the magazine, and you fire. The weapon will chamber the last round but the weapon overlay doesn't update to show an empty magazine.